openassessment/workflow_batch_update_api.py

```diff
@@ -0,0 +1,595 @@
+
+from opaque_keys.edx.keys import UsageKey
+from openassessment.runtime_imports.functions import modulestore
+from openassessment.assessment.models import PeerWorkflow
+from openassessment.workflow import api
+
+logger = logging.getLogger(__name__)
+
+
+@shared_task(bind=True,
+             acks_late=True,
+             autoretry_for=(Exception,),
+             max_retries=5,
+             retry_backoff=True,
+             retry_backoff_max=500,
+             retry_jitter=True)
+def update_workflows_for_all_blocked_submissions_task(self):  # pylint: disable=unused-argument
+    """
+    Async task wrapper
+    """
+    return update_workflows_for_all_blocked_submissions()
+
+
+@shared_task(bind=True,
+             acks_late=True,
+             autoretry_for=(Exception,),
+             max_retries=5,
+             retry_backoff=True,
+             retry_backoff_max=500,
+             retry_jitter=True)
+def update_workflows_for_course_task(self, course_id):  # pylint: disable=unused-argument
+    """
+    Async task wrapper
+    """
+    return update_workflows_for_course(course_id)
+
+
+@shared_task(bind=True,
+             acks_late=True,
+             autoretry_for=(Exception,),
+             max_retries=5,
+             retry_backoff=True,
+             retry_backoff_max=500,
+             retry_jitter=True)
+def update_workflows_for_ora_block_task(self, item_id):  # pylint: disable=unused-argument
+    """
+    Async task wrapper
+    """
+    return update_workflows_for_ora_block(item_id)
+
+
+@shared_task(bind=True,
+             acks_late=True,
+             autoretry_for=(Exception,),
+             max_retries=5,
+             retry_backoff=True,
+             retry_backoff_max=500,
+             retry_jitter=True)
+# pylint: disable=unused-argument
+def update_workflow_for_submission_task(self, submission_uuid, assessment_requirements, course_settings):
+    """
+    Async task wrapper
+    """
+    return update_workflow_for_submission(submission_uuid, assessment_requirements, course_settings)
+
+
+def update_workflows_for_all_blocked_submissions():
+    """
+    Updates ORA workflows for submissions meeting following filtering criteria:
+     - Flexible Peer Grading ON
+     - ungraded submissions that are >7 days old
+
+     Raises:
+        OraWorkflowBatchUpdateException: If batch process fails and cannot continue, e.g. number of errors
+                                        threshold was exceeded, etc.
+    """
+    try:
+        start = time.time()
+
+        peer_workflows = get_blocked_peer_workflows()
+        workflow_update_parameters = get_workflow_update_parameters(peer_workflows)
+        if workflow_update_parameters.get("courses") is not None:
+            for course_object in workflow_update_parameters.get("courses"):
+                # execute asynchronously (submit Celery task)
+                update_workflows_for_course_task.apply_async([course_object["course_id"], workflow_update_parameters])
```

Comment:
This seems clunky, why are we passing the entire `workflow_update_parameters` when we only care about the parameters for a given course at this level?

This would also make unpacking of data at each subsequent level less complicated so you could get the data inline instead of having a bunch of additional utility functions to unpack that data for you.

Reply:
I have, but I think passing around the same, giant data blob is not the most efficient way to do it.

If I'm misunderstanding how the data gets shared to the async task, please correct me, but wouldn't this mean *every* one of those tasks has to be sent the **entire** "cached" lookup for every course, ORA, and submission we care about? Even if the scope only cares about a very limited subset of that data?

Reply:
That could make sense if you had, for example, a statically typed data object to pass around and didn't want to make variations for the different contexts. But in Python land, you're not getting anything by adding extra context unnecessarily to a dictionary. You are just creating extra work in unpacking the data later.

Reply:
Again, this is a compromise after considering different approaches, it is a trade off but comes with advantages (easily scalable, etc). If it proves to be problematic for performance reasons we can update it.

Reply:
I still disagree, on the record, but I don't see any further value arguing over it if you're willing to stand behind that decision.